### PR TITLE
[RFC] context.Context and Tracing

### PR DESCRIPTION
**Pull request by lstoll**
Hi All, Opening this partial PR to request some commentary and discussion around the ideas here.

tl;dr - introducing `context.Context` for 'requests' from AWS, and starting annotations with [OpenTracing](http://opentracing.io/). The former _does not_ require the latter to happen.

The first component is the introduction of [`context.Context`](https://godoc.org/golang.org/x/net/context) to our "Requests". Whilst technically this library doesn't directly initiate requests, the tasks we get from SWF essentially represent this so we should treat them as such. The context isn't really used initially apart from carrying a RequestID, but it gets us the foundations for moving request scoped data and timeouts through these calls. This can be particular useful for activities that make futher remote/rpc calls in to other systems. Because we already have the commonly used `FSMContext`, in places where this is passed a method is provided on it to get the request context. Otherwise a `context.Context` is passed as the first param as is custom.

Secondly, we start to introduce [OpenTracing](http://opentracing.io/), with the context being the method to transport it around. Spans are started/finished where it makes sense (aws request, decision tasks). The span information is passed via the `context.Context`, and a helper is provided on `fsm.Context` as well. This can be used to add spans or events as it makes sense. This gives the library consumer the option to plug in a tracing mechanism, and provide more explicit and context oriented views in to the flow than logging does.

What do people think about the direction this is going in? If there aren't any objections, I'd like to pursue this further.

### CODE REVIEW

**Comment by sclasen** [comment]
Looking good, +1 for not modifying the signature of Decider and attaching to the FSMContext.

Doesn't look like it will cause a ton of pain to bring in 

Thoughts anyone else?

**Comment by amerine** [comment]
This makes me wish we had some benchmarking, but it looks like a most excellent start to me. Would be happy to see this continue.

**Comment by lstoll** [comment]
@sclasen it wasn't too bad, there's a branch `lstoll-swfsm-tracing` on our app with it. The most annoying part was lots of copy and pasting `ctx context.Context` to the beginning of the all activity functions. Deciders were fine tho.

@amerine any particular benchmark concerns here? I might be able to address them directly.

I'm going to break this in to two parts - Firstly will be the addition of `context.Context`. This is the more defined and structural chance, so it'd be nice to get it in and clear.

The tracing stuff will come after - once context is in that's a fairly low impact change. There's also more discussion and consideration needed there as well. Like, do we want to record a span for every time we poll for something? Do we want the span to reflect the poll itself, which can be up to 60s of waiting? (that makes for some pretty crunched up output)
zipkin - traces 2016-07-30 12-21-51. These are all things we can consider independent of the context stuff though.

**Comment by sclasen** [comment]
Thoughts on introducing a activity analog to `FSMContext` and hanging the context off of it?

I had meant to do that here https://github.com/sclasen/swfsm/pull/41   but never got it wrapped up.

**Comment by amerine** [comment]
@lstoll no concerns at all! It would just be nice is all. Its just great seeing how the tweaks impact performance.

**Comment by lstoll** [comment]
I think an `ActivityContext` could make sense. I can see purists objecting,
but I think it keeps us consistent with how we'd be doing things in the
decider and avoid having 40 contexts passed around. There's be one context
object that include everything for the Activity or FSM, where the _Request_
context (and span) are just a sub-component of that.

Let me sketch it out. Would you envisage *swf.PollForActivityTaskOutput
coming under this somehow, or still being it's own parameter? I'd tend to
think the latter.

**Comment by sclasen** [comment]
+1 task still as own param.

**Comment by amerine** [comment]
Should we spruce this up and land it @sclasen & @lstoll?

**Comment by lstoll** [comment]
Closing this out now - it was a nice idea, but I'm really focusing in this area at the moment.
